pkg/azuredisk/azdrivernodeplugin.go

```diff
@@ -0,0 +1,103 @@
+// +build azurediskv2
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azuredisk
+
+import (
+	"context"
+	"time"
+
+	//corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientSet "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	azuredisk "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1alpha1"
+	azVolumeClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+)
+
+
+var partitionLabel  = "azdrivernodes.disk.csi.azure.com/partition"
+
+// RegisterAzDriverNodeOrDie creates custom resource for this driverNode
+func RegisterAzDriverNodeOrDie(NodeID string,
```

Comment:
Is it useful/important to separate the initial creation of the `AzDriverNode` object from the heartbeat loop? A failure to register initially would exit the process and result in a restart of the container which might recover, but we should also handle that case if the failures happen repeatedly sometime after the intial registration. Would restarting the container be the right thing to do? What about operations that are currently in-flight if we restart?

Reply:
Ah, the heartbeat loop would only need to update the `Status` field which can be done independently of the whole. Still, we should consider how to handle a permanent loss of connection to the control plane in the heartbeat loop.

Reply:
Then again, what if a user accidentally deletes the `AzDriverNode` object? Shouldn't the heartbeat loop re-create it?

Reply:
A general approach that is taken here is this:
1. We should not take any traffic until we are able to register. If we are failing to register, I think there is no point starting the gRPC server, because that would have nastier issues.
2. Once we have registered, started gRPC server and started taking traffic, we should not crash the driver node just because we are not able to send heartbeat. In other words RunAzDriverNodeHeartbeatLoop will catch all errors, but we are not going to kill the driver because we failed to communicate. The stale heartbeat would make sure that this node is deprioritized.

That's the reason you see the calls are split, when we register, we are just showing up so that extender doesn't exclude the node at the time of filtering (think when cluster is newly built). Its our way of telling extender that the csi driver runs here.
We start sending ready for allocation==true along with heartbeat only after gRPC server is up.
So extender can decide not to allocate pods to this node by weighing it very low based on readyForAllocation and how old the heartbeat is. extender might ignore readyForAllocation==false, if we don't have too many nodes where driver is running.
let me know what you think.

Reply:
Make sense.